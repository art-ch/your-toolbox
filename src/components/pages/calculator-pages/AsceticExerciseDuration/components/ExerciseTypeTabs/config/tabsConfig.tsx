import { GiMeditation } from 'react-icons/gi';
import { FaPerson, FaPersonRunning, FaPersonSwimming } from 'react-icons/fa6';
import { ExerciseTypeConfig, ThemeColor } from '../ExerciseTypeTabs.types';
import { CalculateCleansingCyclesForm } from '../../forms/dynamic-exercise-forms/CalculateCleansingCyclesForm';
import { CalculateSpeedForm } from '../../forms/dynamic-exercise-forms/CalculateSpeedForm';
import { CalculateTotalTimeForm } from '../../forms/dynamic-exercise-forms/CalculateTotalTimeForm';
import { CalculateDurationFromLayersForm } from '../../forms/static-exercise-forms/CalculateDurationFromLayers';
import { CalculateLayersFromDurationForm } from '../../forms/static-exercise-forms/CalculateLayersFromDuration';
import { CalculateMentalLayersCleansedForm } from '../../forms/static-exercise-forms/CalculateMentalLayersCleansedForm';
import { CalculateTemperatureForm } from '../../forms/static-exercise-forms/CalculateTemperatureForm';
import { CalculateExposureTimeForm } from '../../forms/static-exercise-forms/CalculateExposureTimeForm';

export const EXERCISE_TYPE_TABS_CONFIG: ExerciseTypeConfig = [
  {
    key: 'dynamic',
    themeColor: ThemeColor.RED,
    icon: <FaPersonRunning />,
    forms: [
      CalculateCleansingCyclesForm,
      CalculateSpeedForm,
      CalculateTotalTimeForm
    ]
  },
  {
    key: 'static',
    themeColor: ThemeColor.INDIGO,
    icon: <FaPerson />,
    children: [
      {
        key: 'ground',
        themeColor: ThemeColor.GREEN,
        icon: <GiMeditation />,
        forms: [
          CalculateDurationFromLayersForm,
          CalculateLayersFromDurationForm
        ]
      },
      {
        key: 'water',
        themeColor: ThemeColor.CYAN,
        icon: <FaPersonSwimming />,
        forms: [
          CalculateExposureTimeForm,
          CalculateMentalLayersCleansedForm,
          CalculateTemperatureForm
        ]
      }
    ]
  }
];
