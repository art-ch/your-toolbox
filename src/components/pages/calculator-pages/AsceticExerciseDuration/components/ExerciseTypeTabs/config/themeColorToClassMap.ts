import { ThemeColor } from '../ExerciseTypeTabs.types';

export const THEME_COLOR_TO_CLASS_MAP = {
  [ThemeColor.RED]: {
    text: 'text-red-500',
    activeBg: 'data-[state=active]:bg-red-500'
  },
  [ThemeColor.INDIGO]: {
    text: 'text-indigo-500',
    activeBg: 'data-[state=active]:bg-indigo-500'
  },
  [ThemeColor.GREEN]: {
    text: 'text-green-500',
    activeBg: 'data-[state=active]:bg-green-500'
  },
  [ThemeColor.CYAN]: {
    text: 'text-cyan-500',
    activeBg: 'data-[state=active]:bg-cyan-500'
  }
};
