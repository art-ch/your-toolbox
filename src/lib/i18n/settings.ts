export const languages = ['en', 'ua', 'ru'];
export const defaultLanguage = 'en';
export const defaultNamespace = 'common';

export function getI18nOptions(
  lng = defaultLanguage,
  ns: string | string[] = defaultNamespace
) {
  return {
    supportedLngs: languages,
    fallbackLng: defaultLanguage,
    lng,
    fallbackNS: defaultNamespace,
    defaultNS: defaultNamespace,
    ns,
    interpolation: {
      escapeValue: false
    }
  };
}
