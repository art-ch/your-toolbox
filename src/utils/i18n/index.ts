export * from './applyGrammarCase';
export * from './formatTime';
export * from './utils';
