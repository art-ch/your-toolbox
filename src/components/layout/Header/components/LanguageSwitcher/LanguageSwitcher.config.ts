import { defaultLanguage } from '@/lib/i18n/settings';
import { Language } from '@/lib/i18n/types';

type LanguageSwitcherConfig = Record<Language, { label: string }>;

export const languageSwitcherConfig: LanguageSwitcherConfig = {
  [defaultLanguage]: {
    label: 'English'
  },
  ['ru']: {
    label: 'Русский'
  },
  ['uk']: {
    label: 'Українська'
  }
};
