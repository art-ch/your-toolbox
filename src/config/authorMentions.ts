import { AuthorMentionType } from '@/types/types';
import { TFunction } from 'i18next';
import { Language } from '@/lib/i18n/types';

export const getTorsunovAuthorMention = (
  t: TFunction,
  language: Language
): AuthorMentionType => ({
  name: t('asceticExerciseDuration:author'),
  href: language === 'ru' ? 'https://torsunov.ru' : 'https://torsunov.ru/en'
});
