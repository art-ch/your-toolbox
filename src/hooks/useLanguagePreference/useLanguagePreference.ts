'use client';

import { useState, useEffect } from 'react';
import { useRouter, usePathname } from 'next/navigation';
import { languages } from '@/lib/i18n/settings';
import Cookies from 'js-cookie';

export function useLanguagePreference() {
  const [preferredLanguage, setPreferredLanguage] = useState<string | null>(
    null
  );
  const router = useRouter();
  const pathname = usePathname();

  // Load saved preference on component mount
  useEffect(() => {
    const savedLanguage = Cookies.get('NEXT_LOCALE');
    if (savedLanguage && languages.includes(savedLanguage)) {
      setPreferredLanguage(savedLanguage);
    }
  }, []);

  // Change language and save preference
  const changeLanguage = (newLanguage: string) => {
    if (!languages.includes(newLanguage)) {
      console.error(`Language ${newLanguage} is not supported`);
      return;
    }

    // Set cookie with a reasonable expiration (1 year)
    Cookies.set('NEXT_LOCALE', newLanguage, {
      expires: 365,
      path: '/',
      sameSite: 'lax' // Good security practice
    });

    setPreferredLanguage(newLanguage);

    // Update URL to reflect language change
    const segments = pathname.split('/');
    segments[1] = newLanguage; // Replace the language segment
    const newPath = segments.join('/');

    router.push(newPath);
  };

  return { preferredLanguage, changeLanguage };
}
