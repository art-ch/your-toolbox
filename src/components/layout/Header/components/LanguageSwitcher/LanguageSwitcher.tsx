'use client';

import { languages } from '@/lib/i18n/settings';
import { useLanguagePreference } from '@/hooks/useLanguagePreference/useLanguagePreference';

import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger
} from '@/components/ui/dropdown-menu';

export function LanguageSwitcher({ currentLng }: { currentLng: string }) {
  const { changeLanguage } = useLanguagePreference();

  return (
    <DropdownMenu>
      <DropdownMenuTrigger className="max-w-fit justify-self-end cursor-pointer">
        {currentLng.toUpperCase()}
      </DropdownMenuTrigger>
      <DropdownMenuContent>
        {languages.map((lng) => (
          <DropdownMenuItem
            key={lng}
            className={`${currentLng === lng ? 'font-bold' : ''}`}
            onClick={() => changeLanguage(lng)}
          >
            {lng.toUpperCase()}
          </DropdownMenuItem>
        ))}
      </DropdownMenuContent>
    </DropdownMenu>
  );
}
