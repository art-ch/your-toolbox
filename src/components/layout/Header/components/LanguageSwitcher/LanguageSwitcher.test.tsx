import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import { LanguageSwitcher } from './LanguageSwitcher';
import { useLanguagePreference } from '@/hooks/useLanguagePreference/useLanguagePreference';
import { languageSwitcherConfig } from './LanguageSwitcher.config';
import { languages } from '@/lib/i18n/settings';

// Mock the useLanguagePreference hook
jest.mock('@/hooks/useLanguagePreference/useLanguagePreference', () => ({
  useLanguagePreference: jest.fn()
}));

// Mock the UI dropdown components
jest.mock('@/components/ui/dropdown-menu', () => ({
  DropdownMenu: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="dropdown-menu">{children}</div>
  ),
  DropdownMenuTrigger: ({
    children,
    className
  }: {
    children: React.ReactNode;
    className: string;
  }) => (
    <button data-testid="dropdown-trigger" className={className}>
      {children}
    </button>
  ),
  DropdownMenuContent: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="dropdown-content">{children}</div>
  ),
  DropdownMenuItem: ({
    children,
    className,
    onClick
  }: {
    children: React.ReactNode;
    className: string;
    onClick: () => void;
  }) => (
    <div
      data-testid="dropdown-item"
      role="menuitem"
      className={className}
      onClick={onClick}
    >
      {children}
    </div>
  )
}));

describe('LanguageSwitcher Component', () => {
  const mockChangeLanguage = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useLanguagePreference as jest.Mock).mockReturnValue({
      changeLanguage: mockChangeLanguage
    });
  });

  it('should render with the current language label', () => {
    render(<LanguageSwitcher currentLanguage="en" />);

    expect(
      within(screen.getByTestId('dropdown-trigger')).getByText(
        languageSwitcherConfig.en.label
      )
    ).toBeInTheDocument();
  });

  it('should show dropdown menu with all available languages when clicked', () => {
    render(<LanguageSwitcher currentLanguage="en" />);

    fireEvent.click(
      within(screen.getByTestId('dropdown-trigger')).getByText(
        languageSwitcherConfig.en.label
      )
    );

    languages.forEach((language) => {
      expect(
        within(screen.getByTestId('dropdown-content')).getByText(
          languageSwitcherConfig[language].label
        )
      ).toBeInTheDocument();
    });
  });

  it('should call changeLanguage when a language is selected', () => {
    render(<LanguageSwitcher currentLanguage="en" />);

    fireEvent.click(
      within(screen.getByTestId('dropdown-trigger')).getByText(
        languageSwitcherConfig.en.label
      )
    );

    fireEvent.click(
      within(screen.getByTestId('dropdown-content')).getByText(
        languageSwitcherConfig.ru.label
      )
    );

    expect(mockChangeLanguage).toHaveBeenCalledWith('ru');
  });

  it('should apply bold styling to current language in dropdown', () => {
    render(<LanguageSwitcher currentLanguage="en" />);

    // Open the dropdown
    fireEvent.click(
      within(screen.getByTestId('dropdown-trigger')).getByText(
        languageSwitcherConfig.en.label
      )
    );

    // Get all dropdown items
    const dropdownItems = screen.getAllByTestId('dropdown-item');

    // Find the current language item
    const currentLanguageItem = dropdownItems.find(
      (item) => item.textContent === languageSwitcherConfig.en.label
    );

    // Check if it has the font-bold class
    expect(currentLanguageItem).toHaveClass('font-bold');

    // Check that other language items don't have font-bold
    const otherLanguageItems = dropdownItems.filter(
      (item) => item.textContent !== languageSwitcherConfig.en.label
    );

    otherLanguageItems.forEach((item) => {
      expect(item).not.toHaveClass('font-bold');
    });
  });

  it('should render with a different current language', () => {
    render(<LanguageSwitcher currentLanguage="uk" />);

    expect(
      within(screen.getByTestId('dropdown-trigger')).getByText(
        languageSwitcherConfig.uk.label
      )
    ).toBeInTheDocument();
  });

  it('should allow switching to each available language', () => {
    render(<LanguageSwitcher currentLanguage="en" />);

    // Test switching to each language
    languages.forEach((language) => {
      if (language === 'en') return; // Skip current language

      // Find all dropdown items
      const menuItems = screen.getAllByTestId('dropdown-item');

      // Find and click the language item
      const langItem = menuItems.find(
        (item) => item.textContent === languageSwitcherConfig[language].label
      );

      fireEvent.click(langItem!);

      // Verify the changeLanguage function was called with the correct language
      expect(mockChangeLanguage).toHaveBeenCalledWith(language);

      // Reset for next iteration
      jest.clearAllMocks();
    });
  });

  it('should match a snapshot', () => {
    const { container } = render(<LanguageSwitcher currentLanguage="en" />);

    expect(container).toMatchSnapshot();
  });
});
