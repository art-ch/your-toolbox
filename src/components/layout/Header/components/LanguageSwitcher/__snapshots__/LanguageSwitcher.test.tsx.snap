// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher Component should match a snapshot 1`] = `
<div>
  <div
    data-testid="dropdown-menu"
  >
    <button
      class="max-w-fit justify-self-end cursor-pointer"
      data-testid="dropdown-trigger"
    >
      English
    </button>
    <div
      data-testid="dropdown-content"
    >
      <div
        class="font-bold"
        data-testid="dropdown-item"
        role="menuitem"
      >
        English
      </div>
      <div
        class=""
        data-testid="dropdown-item"
        role="menuitem"
      >
        Українська
      </div>
      <div
        class=""
        data-testid="dropdown-item"
        role="menuitem"
      >
        Русский
      </div>
    </div>
  </div>
</div>
`;
